### DISCUSSION TITLE
Multiple authentication providers

### DISCUSSION THREAD

**Discussion by pwoji** [JonPSmith/AuthPermissions.AspNetCore#66] [Q&A] [answered]
Does this solution work if I have both Azure AD and Azure AD B2C configured as authentication providers?

**Comment by JonPSmith** [comment, marked as answer]
hi @pwojee,

ASP.NET Core allows multiple authentication and you can intercept the OpenId `OnTokenValidated` event (see [AzureAdOpenIdExtension](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions.AspNetCore/OpenIdCode/AzureAdOpenIdExtension.cs) of your two authentication providers to link the AuthP user.

AuthP provides a way to register a **single** Azure AD and sets up the `OnTokenValidated` event code, but you can manually add another by adding the [AzureAdOpenIdExtension](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions.AspNetCore/OpenIdCode/AzureAdOpenIdExtension.cs) to the `OnTokenValidated` event of the second Azure AD.

AuthP relies on the userId of a user. There are various ways to set up a AuthP user - I suggest you look at the article [Three ways to securely add new users to an application using the AuthP library](https://www.thereformedprogrammer.net/three-ways-to-securely-add-new-users-to-an-application-using-the-authp-library/) for an overview and then the [AuthP documentation](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki) for the details.

**Reply by pwoji** [reply]
Thank you.
